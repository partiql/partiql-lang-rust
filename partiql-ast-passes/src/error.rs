use partiql_catalog::call_defs::CallLookupError;
use thiserror::Error;

/// Contains the errors that occur during AST transformations
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct AstTransformationError {
    pub errors: Vec<AstTransformError>,
}

/// Represents an AST transform Error
#[derive(Error, Debug, Clone, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum AstTransformError {
    /// Indicates that AST lowering has not yet been implemented for this feature.
    #[error("Not yet implemented: {0}")]
    NotYetImplemented(String),

    /// Indicates that there is an internal error that was not due to user input or API violation.
    #[error("Illegal State: {0}")]
    IllegalState(String),

    /// Indicates that there was an error interpreting a literal value.
    #[error("Error with literal: {literal}: {error}")]
    Literal { literal: String, error: String },

    /// Invalid number of arguments to the function call.
    #[error("Invalid number of arguments: {0}")]
    InvalidNumberOfArguments(String),

    /// Indicates that this function is not supported.
    #[error("Unsupported function: {0}")]
    UnsupportedFunction(String),

    /// Indicates that this aggregation function is not supported.
    #[error("Unsupported aggregation function: {0}")]
    UnsupportedAggregationFunction(String),

    /// Any other lowering error.
    #[error("Lowering error: {0}")]
    Unknown(String),

    /// Indicates that a `HAVING` clause was provided without a `GROUP BY`
    #[error("HAVING clause provided without GROUP BY")]
    HavingWithoutGroupBy,
}

impl From<CallLookupError> for AstTransformError {
    fn from(err: CallLookupError) -> Self {
        match err {
            CallLookupError::InvalidNumberOfArguments(e) => {
                AstTransformError::InvalidNumberOfArguments(e)
            }
            e => AstTransformError::Unknown(e.to_string()),
        }
    }
}
