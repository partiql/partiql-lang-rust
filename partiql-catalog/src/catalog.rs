use crate::call_defs::ScalarFnCallSpecs;
use crate::scalar_fn::ScalarFunction;
use crate::table_fn::TableFunction;
use partiql_common::catalog::{CatalogId, EntryId, ObjectId};
use partiql_types::PartiqlShape;
use std::collections::HashMap;
use std::fmt::Debug;
use std::sync::atomic::{AtomicU64, Ordering};
use thiserror::Error;
use unicase::UniCase;

/// Contains the errors that occur during Catalog related operations
#[derive(Error, Debug, Clone, PartialEq)]
#[error("Catalog error: encountered errors")]
pub struct CatalogError {
    pub errors: Vec<CatalogErrorKind>,
}

impl CatalogError {
    #[must_use]
    pub fn new(errors: Vec<CatalogErrorKind>) -> Self {
        CatalogError { errors }
    }
}

/// Catalog Error kind
///
/// ### Notes
/// This is marked `#[non_exhaustive]`, to reserve the right to add more variants in the future.
#[derive(Error, Debug, Clone, PartialEq)]
#[non_exhaustive]
pub enum CatalogErrorKind {
    /// Entry exists error.
    #[error("Catalog error: entry already exists for `{0}`")]
    EntryExists(String),

    /// Entry error.
    #[error("Catalog error: `{0}`")]
    EntryError(String),

    /// Any other catalog error.
    #[error("Catalog error: unknown error")]
    Unknown,
}

pub trait Catalog: Debug {
    fn add_table_function(&mut self, info: TableFunction) -> Result<ObjectId, CatalogError>;
    fn add_scalar_function(&mut self, info: ScalarFunction) -> Result<ObjectId, CatalogError>;

    fn add_type_entry(&mut self, entry: TypeEnvEntry<'_>) -> Result<ObjectId, CatalogError>;

    fn get_function(&self, name: &str) -> Option<FunctionEntry<'_>>;

    fn get_function_by_id(&self, id: ObjectId) -> Option<FunctionEntry<'_>>;

    fn resolve_type(&self, name: &str) -> Option<TypeEntry>;
}

#[derive(Debug)]
pub struct TypeEnvEntry<'a> {
    name: UniCase<String>,
    aliases: Vec<&'a str>,
    ty: PartiqlShape,
}

impl<'a> TypeEnvEntry<'a> {
    #[must_use]
    pub fn new(name: &str, aliases: &[&'a str], ty: PartiqlShape) -> Self {
        TypeEnvEntry {
            name: UniCase::from(name.to_string()),
            aliases: aliases.to_vec(),
            ty,
        }
    }
}

#[derive(Debug, Clone)]
pub struct TypeEntry {
    id: ObjectId,
    ty: PartiqlShape,
}

impl TypeEntry {
    #[must_use]
    pub fn id(&self) -> &ObjectId {
        &self.id
    }

    #[must_use]
    pub fn ty(&self) -> &PartiqlShape {
        &self.ty
    }
}

#[derive(Debug)]
#[allow(dead_code)]
pub struct FunctionEntry<'a> {
    id: ObjectId,
    function: &'a FunctionEntryFunction,
}

#[derive(Debug)]
pub enum FunctionEntryFunction {
    Table(TableFunction),
    Scalar(ScalarFnCallSpecs),
    Aggregate(),
}

impl<'a> FunctionEntry<'a> {
    pub fn id(&self) -> &ObjectId {
        &self.id
    }

    #[must_use]
    pub fn entry(&'a self) -> &'a FunctionEntryFunction {
        self.function
    }
}

#[derive(Debug)]
pub struct PartiqlCatalog {
    functions: CatalogEntrySet<FunctionEntryFunction>,
    types: CatalogEntrySet<PartiqlShape>,
    id: CatalogId,
}

impl Default for PartiqlCatalog {
    fn default() -> Self {
        PartiqlCatalog {
            functions: Default::default(),
            types: Default::default(),
            id: 1.into(),
        }
    }
}

impl PartiqlCatalog {}

impl Catalog for PartiqlCatalog {
    fn add_table_function(&mut self, info: TableFunction) -> Result<ObjectId, CatalogError> {
        let call_def = info.call_def();
        let names = call_def.names.clone();
        if let Some((name, aliases)) = names.split_first() {
            let eid = self
                .functions
                .add(name, aliases, FunctionEntryFunction::Table(info))?;
            Ok((self.id, eid).into())
        } else {
            Err(CatalogError::new(vec![CatalogErrorKind::EntryError(
                "Function definition has no name".into(),
            )]))
        }
    }

    fn add_scalar_function(&mut self, info: ScalarFunction) -> Result<ObjectId, CatalogError> {
        let id = self.id;
        let call_def = info.into_call_def();
        let names = call_def.names;
        if let Some((name, aliases)) = names.split_first() {
            self.functions
                .add(
                    name,
                    aliases,
                    FunctionEntryFunction::Scalar(call_def.overloads),
                )
                .map(|eid| ObjectId::new(id, eid))
        } else {
            Err(CatalogError::new(vec![CatalogErrorKind::EntryError(
                "Function definition has no name".into(),
            )]))
        }
    }

    fn add_type_entry(&mut self, entry: TypeEnvEntry<'_>) -> Result<ObjectId, CatalogError> {
        let eid = self
            .types
            .add(entry.name.as_ref(), entry.aliases.as_slice(), entry.ty);

        match eid {
            Ok(eid) => Ok((self.id, eid).into()),
            Err(e) => Err(e),
        }
    }

    fn get_function(&self, name: &str) -> Option<FunctionEntry<'_>> {
        self.functions
            .find_by_name(name)
            .map(|(e, f)| self.to_function_entry(e, f))
    }

    fn get_function_by_id(&self, id: ObjectId) -> Option<FunctionEntry<'_>> {
        assert_eq!(self.id, id.catalog_id());
        self.functions
            .find_by_id(&id.entry_id())
            .map(|(e, f)| self.to_function_entry(e, f))
    }

    fn resolve_type(&self, name: &str) -> Option<TypeEntry> {
        self.types.find_by_name(name).map(|(eid, entry)| TypeEntry {
            id: (self.id, eid).into(),
            ty: entry.clone(),
        })
    }
}

impl PartiqlCatalog {
    fn to_function_entry<'a>(
        &'a self,
        eid: EntryId,
        entry: &'a FunctionEntryFunction,
    ) -> FunctionEntry<'a> {
        FunctionEntry {
            id: (self.id, eid).into(),
            function: entry,
        }
    }
}

#[derive(Debug)]
struct CatalogEntrySet<T> {
    entries: HashMap<EntryId, T>,
    by_name: HashMap<UniCase<String>, EntryId>,
    by_alias: HashMap<UniCase<String>, EntryId>,

    next_id: AtomicU64,
}

impl<T> Default for CatalogEntrySet<T> {
    fn default() -> Self {
        CatalogEntrySet {
            entries: Default::default(),
            by_name: Default::default(),
            by_alias: Default::default(),
            next_id: 1.into(),
        }
    }
}

impl<T> CatalogEntrySet<T> {
    fn add(&mut self, name: &str, aliases: &[&str], info: T) -> Result<EntryId, CatalogError> {
        let mut errors = vec![];
        let name = UniCase::from(name);
        let aliases: Vec<UniCase<String>> = aliases
            .iter()
            .map(|a| UniCase::from((*a).to_string()))
            .collect();

        for a in &aliases {
            if self.by_alias.contains_key(a) {
                errors.push(CatalogErrorKind::EntryExists(a.as_ref().to_string()));
            }
        }

        if self.by_name.contains_key(&name) {
            errors.push(CatalogErrorKind::EntryExists(name.to_string()));
        }

        let id = self.next_id.fetch_add(1, Ordering::SeqCst).into();

        if let Some(_old_val) = self.entries.insert(id, info) {
            errors.push(CatalogErrorKind::Unknown);
        }

        match errors.is_empty() {
            true => {
                self.by_name.insert(name, id);

                for a in aliases {
                    self.by_alias.insert(a, id);
                }

                Ok(id)
            }
            _ => Err(CatalogError::new(errors)),
        }
    }

    fn find_by_id(&self, eid: &EntryId) -> Option<(EntryId, &T)> {
        self.entries.get(eid).map(|e| (*eid, e))
    }

    fn find_by_name(&self, name: &str) -> Option<(EntryId, &T)> {
        let name = UniCase::from(name);
        let eid = self.by_name.get(&name).or(self.by_alias.get(&name));

        eid.and_then(|eid| self.find_by_id(eid))
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn todo() {}
}
