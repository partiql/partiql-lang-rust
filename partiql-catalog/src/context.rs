use partiql_value::datum::RefTupleView;
use partiql_value::{BindingsName, DateTime, Tuple, Value};
use std::any::Any;
use std::borrow::Cow;
use std::fmt::Debug;

pub trait Bindings<T>: Debug
where
    T: Clone + Debug,
{
    fn get<'a>(&'a self, name: &BindingsName<'_>) -> Option<Cow<'a, T>>;
}

impl Bindings<Value> for Tuple {
    fn get<'a>(&'a self, name: &BindingsName<'_>) -> Option<Cow<'a, Value>> {
        self.get(name).map(Cow::Borrowed)
    }
}

impl<T> Bindings<Value> for &T
where
    for<'a> T: RefTupleView<'a, Value>,
{
    fn get<'a>(&'a self, name: &BindingsName<'_>) -> Option<Cow<'a, Value>> {
        self.get_val(name)
    }
}

#[derive(Debug)]
pub struct SystemContext {
    pub now: DateTime,
}

/// Represents a session context that is used during evaluation of a plan.
pub trait SessionContext: Debug {
    fn system_context(&self) -> &SystemContext;

    fn user_context(&self, name: &str) -> Option<&(dyn Any)>;
}
