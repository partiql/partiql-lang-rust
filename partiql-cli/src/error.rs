use miette::{Diagnostic, LabeledSpan, SourceCode};
use partiql_parser::{ParseError, ParserError};
use partiql_source_map::location::{BytePosition, Location};

use thiserror::Error;

#[derive(Debug, Error, Diagnostic)]
#[error("Error for query `{query}`")]
pub struct CLIErrors {
    query: String,
    #[related]
    related: Vec<CLIError>,
}

impl CLIErrors {
    pub fn from_parser_error(err: ParserError) -> Self {
        let query = err.text.to_string();

        let related = err
            .errors
            .into_iter()
            .map(|e| CLIError::from_parse_error(e, &query))
            .collect();
        CLIErrors { query, related }
    }
}

#[derive(Debug, Error)]
pub enum CLIError {
    #[error("PartiQL syntax error:")]
    SyntaxError {
        src: String,
        msg: String,
        loc: Location<BytePosition>,
    },
    #[error("Internal Compiler Error - please report this (https://github.com/partiql/partiql-lang-rust/issues).")]
    InternalCompilerError { src: String },
}

impl Diagnostic for CLIError {
    fn source_code(&self) -> Option<&dyn SourceCode> {
        match self {
            CLIError::SyntaxError { src, .. } => Some(src),
            CLIError::InternalCompilerError { src, .. } => Some(src),
        }
    }

    fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
        match self {
            CLIError::SyntaxError { msg, loc, .. } => {
                Some(Box::new(std::iter::once(LabeledSpan::new(
                    Some(msg.to_string()),
                    loc.start.0 .0 as usize,
                    loc.end.0 .0 as usize - loc.start.0 .0 as usize,
                ))))
            }
            CLIError::InternalCompilerError { .. } => None,
        }
    }
}

impl CLIError {
    pub fn from_parse_error(err: ParseError, source: &str) -> Self {
        match err {
            ParseError::SyntaxError(partiql_source_map::location::Located { inner, location }) => {
                CLIError::SyntaxError {
                    src: source.to_string(),
                    msg: format!("Syntax error `{}`", inner),
                    loc: location,
                }
            }
            ParseError::UnexpectedToken(partiql_source_map::location::Located {
                inner,
                location,
            }) => CLIError::SyntaxError {
                src: source.to_string(),
                msg: format!("Unexpected token `{}`", inner.token),
                loc: location,
            },
            ParseError::LexicalError(partiql_source_map::location::Located { inner, location }) => {
                CLIError::SyntaxError {
                    src: source.to_string(),
                    msg: format!("Lexical error `{}`", inner),
                    loc: location,
                }
            }
            ParseError::Unknown(location) => CLIError::SyntaxError {
                src: source.to_string(),
                msg: "Unknown parser error".to_string(),
                loc: Location {
                    start: location,
                    end: location,
                },
            },
            ParseError::IllegalState(_location) => CLIError::InternalCompilerError {
                src: source.to_string(),
            },
            _ => {
                todo!("Not yet handled {:?}", err);
            }
        }
    }
}
