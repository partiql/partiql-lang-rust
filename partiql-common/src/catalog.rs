#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Hash)]
pub struct CatalogId(u64);

impl From<u64> for CatalogId {
    fn from(value: u64) -> Self {
        CatalogId(value)
    }
}

#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Hash)]
pub struct EntryId(u64);

impl From<u64> for EntryId {
    fn from(value: u64) -> Self {
        EntryId(value)
    }
}

#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Hash)]
pub struct ObjectId {
    catalog_id: CatalogId,
    entry_id: EntryId,
}

impl ObjectId {
    pub fn new(catalog_id: CatalogId, entry_id: EntryId) -> Self {
        Self {
            catalog_id,
            entry_id,
        }
    }

    pub fn catalog_id(&self) -> CatalogId {
        self.catalog_id
    }
    pub fn entry_id(&self) -> EntryId {
        self.entry_id
    }
}

impl From<(CatalogId, EntryId)> for ObjectId {
    fn from((catalog_id, entry_id): (CatalogId, EntryId)) -> Self {
        ObjectId::new(catalog_id, entry_id)
    }
}
