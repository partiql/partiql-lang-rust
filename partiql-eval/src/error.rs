use crate::eval::evaluable::Evaluable;
use crate::eval::expr::EvalExpr;
use crate::eval::EvalContext;
use partiql_value::{Tuple, Value};
use std::borrow::Cow;
use thiserror::Error;

/// All errors that occurred during [`partiql_logical::LogicalPlan`] to [`eval::EvalPlan`] creation.
#[derive(Debug)]
pub struct PlanErr {
    pub errors: Vec<PlanningError>,
}

/// An error that can happen during [`partiql_logical::LogicalPlan`] to [`eval::EvalPlan`] creation.
#[derive(Error, Debug, Clone, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum PlanningError {
    /// Feature has not yet been implemented.
    #[error("Not yet implemented: {0}")]
    NotYetImplemented(String),
    /// Internal error that was not due to user input or API violation.
    #[error("Illegal State: {0}")]
    IllegalState(String),
}

/// All errors that occurred during evaluation.
#[derive(Debug)]
pub struct EvalErr {
    pub errors: Vec<EvaluationError>,
}

/// An error that can happen during evaluation.
#[derive(Error, Debug, Clone, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum EvaluationError {
    /// Malformed evaluation plan with graph containing cycle.
    #[error("Evaluation Error: invalid evaluation plan detected `{0}`")]
    InvalidEvaluationPlan(String),
}

/// Used when an error occurs during the the logical to eval plan conversion. Allows the conversion
/// to continue in order to report multiple errors.
#[derive(Debug)]
pub(crate) struct ErrorNode {}

impl ErrorNode {
    pub(crate) fn new() -> ErrorNode {
        ErrorNode {}
    }
}

impl Evaluable for ErrorNode {
    fn evaluate(&mut self, _ctx: &dyn EvalContext) -> Option<Value> {
        panic!("ErrorNode will not be evaluated")
    }

    fn update_input(&mut self, _input: Value, _branch_num: u8) {
        panic!("ErrorNode will not be evaluated")
    }
}

impl EvalExpr for ErrorNode {
    fn evaluate<'a>(&'a self, _bindings: &'a Tuple, _ctx: &'a dyn EvalContext) -> Cow<'a, Value> {
        panic!("ErrorNode will not be evaluated")
    }
}
