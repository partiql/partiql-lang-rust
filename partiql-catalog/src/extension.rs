use crate::catalog::Catalog;
use std::error::Error;
use std::fmt::Debug;
use thiserror::Error;

pub trait Extension: Debug {
    fn name(&self) -> String;

    fn load(&self, catalog: &mut dyn Catalog) -> Result<(), ExtensionResultError>;
}

pub type ExtensionError = Box<dyn Error>;

/// Errors in extension.
///
/// ### Notes
/// This is marked `#[non_exhaustive]`, to reserve the right to add more variants in the future.
#[derive(Error, Debug)]
#[non_exhaustive]
pub enum ExtensionResultError {
    /// Load error. Error loading the extension.
    #[error("Scan error: `{}`", .0)]
    LoadError(ExtensionError),

    /// Read error. This will generally terminate the read operation.
    #[error("Scan error: `{}`", .0)]
    ReadError(ExtensionError),

    /// Data error. Generally this will result in a `MISSING` in place of this data item.
    #[error("Data error: `{}`", .0)]
    DataError(ExtensionError),
}
