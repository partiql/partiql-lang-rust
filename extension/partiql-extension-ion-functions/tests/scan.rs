use partiql_catalog::catalog::{Catalog, PartiqlCatalog};
use partiql_catalog::context::SystemContext;
use partiql_catalog::extension::Extension;
use partiql_eval::env::basic::MapBindings;
use partiql_eval::error::EvaluationError;
use partiql_eval::eval::BasicContext;
use partiql_eval::plan::EvaluationMode;
use partiql_extension_ion_functions::IonExtension;
use partiql_parser::{Parsed, ParserResult};
use partiql_value::{bag, tuple, DateTime, Value};
use std::path::PathBuf;

#[track_caller]
#[inline]
pub(crate) fn parse(statement: &str) -> ParserResult<'_> {
    partiql_parser::Parser::default().parse(statement)
}

#[track_caller]
#[inline]
pub(crate) fn lower(
    catalog: &dyn Catalog,
    parsed: &Parsed<'_>,
) -> partiql_logical::LogicalPlan<partiql_logical::BindingsOp> {
    let planner = partiql_logical_planner::LogicalPlanner::new(catalog);
    planner.lower(parsed).expect("lower")
}

#[track_caller]
#[inline]
pub(crate) fn evaluate(
    catalog: &dyn Catalog,
    logical: partiql_logical::LogicalPlan<partiql_logical::BindingsOp>,
    bindings: MapBindings<Value>,
) -> (Value, Vec<EvaluationError>) {
    let mut planner =
        partiql_eval::plan::EvaluatorPlanner::new(EvaluationMode::Permissive, catalog);

    let mut plan = planner.compile(&logical).expect("Expect no plan error");

    let sys = SystemContext {
        now: DateTime::from_system_now_utc(),
    };
    let ctx = BasicContext::new(bindings, sys);
    let value = if let Ok(out) = plan.execute_mut(&ctx) {
        out.result
    } else {
        Value::Missing
    };
    let errs = ctx.errors.take();
    (value, errs)
}

#[inline]
#[allow(dead_code)]
pub(crate) fn evaluate_with_ion_scan(
    statement: &str,
    env: &Option<Value>,
) -> (Value, Vec<EvaluationError>) {
    let mut catalog = PartiqlCatalog::default();
    let ext = IonExtension {};
    ext.load(&mut catalog)
        .expect("ion extension load to succeed");

    let parsed = parse(statement);
    let lowered = lower(&catalog, &parsed.expect("parse"));
    let bindings = env
        .as_ref()
        .map(std::convert::Into::into)
        .unwrap_or_default();
    evaluate(&catalog, lowered, bindings)
}

#[inline]
#[allow(dead_code)]
pub(crate) fn pass_eval(statement: &str, env: &Option<Value>, expected: &Value) {
    let (out, errs) = evaluate_with_ion_scan(statement, env);

    assert!(out.is_bag());
    assert!(errs.is_empty());
    assert_eq!(&out, expected);
}

fn ion_read_select_distinct(file: &str) {
    let value = bag![
        tuple![("Program", "p1"), ("Operation", "get")],
        tuple![("Program", "p1"), ("Operation", "put")],
        tuple![("Program", "p2"), ("Operation", "get")],
        tuple![("Program", "p2"), ("Operation", "put")],
        tuple![("Program", "p3"), ("Operation", "update")],
    ]
    .into();
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("resources/test");
    path.push(file);
    let path = path.as_path().display();

    let query = format!("SELECT DISTINCT Program, Operation from read_ion('{path}') as fel");
    pass_eval(&query, &None, &value);
}

#[macro_export]
macro_rules! ion {
    ($x:expr) => {
        partiql_extension_ion::boxed_ion::BoxedIonType {}
            .value_from_str($x)
            .expect("boxed ion construct")
            .into_value()
    };
}

fn ion_scan_select_distinct(file: &str) {
    let value = bag![
        tuple![("Program", ion!("\"p1\"")), ("Operation", ion!("\"get\""))],
        tuple![("Program", ion!("\"p1\"")), ("Operation", ion!("\"put\""))],
        tuple![("Program", ion!("\"p2\"")), ("Operation", ion!("\"get\""))],
        tuple![("Program", ion!("\"p2\"")), ("Operation", ion!("\"put\""))],
        tuple![
            ("Program", ion!("\"p3\"")),
            ("Operation", ion!("\"update\""))
        ],
    ]
    .into();
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("resources/test");
    path.push(file);
    let path = path.as_path().display();

    let query = format!("SELECT DISTINCT Program, Operation from scan_ion('{path}') as fel");
    pass_eval(&query, &None, &value);
}

fn ion_scan_range_over(file: &str, query_item: &str) {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("resources/test");
    path.push(file);
    let path = path.as_path().display();

    let query = format!("SELECT {query_item} from scan_ion('{path}') as ion_data");
    let (result, errs) = evaluate_with_ion_scan(&query, &None);

    insta::assert_snapshot!(file, result);
    insta::assert_debug_snapshot!(format!("{file}.errors"), errs);
}

#[test]
fn custom_ion_read_text() {
    ion_read_select_distinct("test.ion");
}

#[test]
fn custom_ion_read_binary() {
    ion_read_select_distinct("test.10n");
}

#[test]
fn custom_ion_read_zstd() {
    ion_read_select_distinct("test.10n.zst");
}

#[test]
fn custom_ion_scan_text() {
    ion_scan_select_distinct("test.ion");
}

#[test]
fn custom_ion_scan_binary() {
    ion_scan_select_distinct("test.10n");
}

#[test]
fn custom_ion_scan_zstd() {
    ion_scan_select_distinct("test.10n.zst");
}

#[test]
fn custom_ion_passthrough() {
    ion_scan_range_over("ion_passthrough_test.ion", "ion_data");
}

#[test]
fn custom_ion_passthrough_structs_text() {
    ion_scan_range_over("ion_passthrough_structs_test.ion", "ion_data.data");
}

#[test]
fn custom_ion_passthrough_bad_text() {
    ion_scan_range_over("ion_passthrough_test.bad.ion", "ion_data.data");
}
