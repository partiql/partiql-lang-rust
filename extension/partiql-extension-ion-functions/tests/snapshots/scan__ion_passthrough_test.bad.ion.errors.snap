---
source: extension/partiql-extension-ion-functions/tests/scan.rs
expression: errs
---
[
    ExtensionResultError(
        ReadError(
            IonReadError(
                Decoding(
                    DecodingError {
                        description: "failed to parse a struct: Decoding(DecodingError { description: \"failed to parse an s-expression: Decoding(DecodingError { description: \\\"invalid Ion syntax encountered\\\\n        offset=99\\\\n        buffer head=<}>\\\\n        buffer tail=<}>\\\\n        buffer len=1\\\\n        \\\", position: Some(Position { byte_offset: 99, byte_length: Some(1), line_column: None }) })\\n        offset=84\\n        buffer head=<(sym::var 1 2 3}>\\n        buffer tail=<(sym::var 1 2 3}>\\n        buffer len=16\\n        \", position: Some(Position { byte_offset: 84, byte_length: Some(16), line_column: None }) })\n        offset=77\n        buffer head=<{data: (sym::var 1 2 3}>\n        buffer tail=<{data: (sym::var 1 2 3}>\n        buffer len=23\n        ",
                        position: Some(
                            Position {
                                byte_offset: 77,
                                byte_length: Some(
                                    23,
                                ),
                                line_column: None,
                            },
                        ),
                    },
                ),
            ),
        ),
    ),
]
